extmap.go

```diff
@@ -0,0 +1,88 @@
+package sdp
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/pions/webrtc/pkg/rtcerr"
+
+	"github.com/pkg/errors"
+)
+
+//ExtMap represents the activation of a single RTP header extension
+type ExtMap struct {
+	Value     int
+	Direction Direction
+	URI       *url.URL
+	ExtAttr   *string
+}
+
+//Clone converts this object to an Attribute
+func (e *ExtMap) Clone() Attribute {
+	return Attribute{Key: "extmap", Value: e.string()}
+}
+
+//Unmarshal creates an Extmap from a string
+func (e *ExtMap) Unmarshal(raw string) error {
+	parts := strings.SplitN(raw, ":", 2)
+	if len(parts) != 2 {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", raw)}, pkgName)
+	}
+
+	fields := strings.Fields(parts[1])
+	if len(parts) < 2 {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", raw)}, pkgName)
+	}
+
+	valdir := strings.Split(fields[0], "/")
+	value, err := strconv.ParseInt(valdir[0], 10, 64)
+	if err != nil {
+		return errors.Wrap(&rtcerr.SyntaxError{Err: fmt.Errorf("%v", valdir[0])}, pkgName)
+	}
+
+	var direction Direction
+	if len(valdir) == 2 {
+		direction = NewDirection(valdir[1])
+	}
+
+	uri, err := url.Parse(fields[1])
+	if err != nil {
+		return errors.Wrap(&rtcerr.UnknownError{Err: err}, pkgName)
+	}
+
+	if len(fields) == 3 {
+		tmp := fields[2]
+		e.ExtAttr = &tmp
+	}
+
+	e.Value = int(value)
+	e.Direction = direction
+	e.URI = uri
+	return nil
+}
+
+//Marshal creates a string from an ExtMap
+func (e *ExtMap) Marshal() string {
+	return attributeKey + e.Name() + ":" + e.string() + endline
+}
+
+func (e *ExtMap) string() string {
+	output := fmt.Sprintf("%d", e.Value)
+	if e.Direction != Direction(unknown) {
```

Comment:
If I don't see the constructor and mistakenly construct an extmap this way:

```
Extmap {
  Direction: Direction(999),
  URI: uri, 
}
```

It will marshal like this:
```
999/unknown http://example.com
```

Reply:
@maxhawkins @wdouglass anyway we can make it easy for people to add special test cases into the example list? Or maybe even make it a requirement for others to submit PRs with special cases that do not pass? Unfortunately not everyone use every possible variation of these attributes all the time and know all the special cases. I many scenarios I found it even hard to find proper examples in the RFCs. Just an idea, not sure if you guys agree.

Reply:
fixed, now it'll marshal like this
`0 http://example.com`

which would be invalid, because 0 isn't a valid value. should Marshal fail in this case?

Reply:
I think Marshal should never fail under the assumption that the document was Unmarshaled first and was not modified by the user manually in the process.

"If marshal failed, it is your fault user" - costa